fence/job/visa_update_cronjob.py

```diff
@@ -0,0 +1,170 @@
+            for j in range(self.n_workers)
+        ]
+        updaters = [
+            loop.create_task(self.updater(i, updater_queue, db_session))
+            for i in range(self.concurrency)
+        ]
+
+        await asyncio.gather(*producers)
+        self.logger.info("Producers done producing")
+        await queue.join()
+
+        await asyncio.gather(*workers)
+        await updater_queue.join()  # blocks until everything in updater_queue is complete
+
+        for u in updaters:
+            u.cancel()
+
+        self.logger.info(
+            "Visa cron job completed in {}".format(
+                datetime.timedelta(seconds=time.time() - start_time)
+            )
+        )
+
+    async def get_user_from_db(self, db_session, chunk_idx):
+        """
+        Window function to get chunks of data from the table
+        """
+        start, stop = self.chunk_size * chunk_idx, self.chunk_size * (chunk_idx + 1)
+        users = db_session.query(User).slice(start, stop).all()
+        return users
+
+    async def producer(self, db_session, queue, chunk_idx):
+        """
+        Produces users from db and puts them in a queue for processing
+        """
+        chunk_size = self.chunk_size
+        while True:
+            users = await self.get_user_from_db(db_session, chunk_idx)
+
+            if users == None:
+                break
+            for user in users:
+                self.logger.info("Producer producing user {}".format(user.username))
+                await queue.put(user)
+            if len(users) < chunk_size:
+                break
+            chunk_idx += 1
+
+    async def worker(self, name, queue, updater_queue):
+        """
+        Create tasks to pass to updater to update visas AND pass updated visas to _verify_jwt_token for verification
+        """
+        while not queue.empty():
+            user = await queue.get()
+            await updater_queue.put(user)
+            queue.task_done()
+
+    async def updater(self, name, updater_queue, db_session):
+        """
+        Update visas in the updater_queue
+        """
+        while True:
+            user = await updater_queue.get()
+            if user.ga4gh_visas_v1:
+                for visa in user.ga4gh_visas_v1:
+                    client = self._pick_client(visa)
+                    self.logger.info(
+                        "Updater {} updating visa for user {}".format(
+                            name, user.username
+                        )
+                    )
+                    client.update_user_visas(user, db_session)
+            else:
+                self.logger.info(
+                    "User {} doesnt have visa. Skipping . . .".format(user.username)
+                )
+
+            updater_queue.task_done()
+
+    def _pick_client(self, visa):
+        """
+        Pick oidc client according to the visa provider
+        """
+        if "ras" in visa.type:
+            return self.ras_client
```

Comment:
```suggestion
        client = None
        if "ras" in visa.type:
            client = self.ras_client
        if not client:
            # raise Exception or write error log - if just a log, then the caller needs to handle a None return value
        return client
```
handles the case where RAS is not configured, and the case where we don't know how to handle this visa type